File: src/features/settings/components/data-settings/agentPresets.ts
export type AgentPreset = {
  id: string;
  name: string;
  defaultEnabled: boolean;
  rootDirSuffix: string;
  ruleFile: string;
};

export const AGENT_PRESETS: AgentPreset[] = [
  {
    id: "codex",
    name: "Codex",
    defaultEnabled: true,
    rootDirSuffix: ".codex",
    ruleFile: "AGENTS.md",
  },
  {
    id: "claude",
    name: "Claude Codex",
    defaultEnabled: true,
    rootDirSuffix: ".claude",
    ruleFile: "CLAUDE.md",
  },
  {
    id: "gemini",
    name: "Gemini",
    defaultEnabled: true,
    rootDirSuffix: ".gemini",
    ruleFile: "AGENTS.md",
  },
  {
    id: "copilot",
    name: "GitHub Copilot",
    defaultEnabled: false,
    rootDirSuffix: ".copilot",
    ruleFile: "AGENTS.md",
  },
  {
    id: "cursor",
    name: "Cursor",
    defaultEnabled: false,
    rootDirSuffix: ".cursor",
    ruleFile: "AGENTS.md",
  },
  {
    id: "windsurf",
    name: "Windsurf",
    defaultEnabled: false,
    rootDirSuffix: ".codeium/windsurf",
    ruleFile: "AGENTS.md",
  },
  {
    id: "kiro",
    name: "Kiro",
    defaultEnabled: false,
    rootDirSuffix: ".kiro",
    ruleFile: "AGENTS.md",
  },
  {
    id: "trae",
    name: "Trae",
    defaultEnabled: false,
    rootDirSuffix: ".trae",
    ruleFile: "AGENTS.md",
  },
  {
    id: "opencode",
    name: "OpenCode",
    defaultEnabled: false,
    rootDirSuffix: ".config/opencode",
    ruleFile: "AGENTS.md",
  },
  {
    id: "roo",
    name: "Roo Code",
    defaultEnabled: false,
    rootDirSuffix: ".roo",
    ruleFile: "AGENTS.md",
  },
  {
    id: "amp",
    name: "Amp",
    defaultEnabled: false,
    rootDirSuffix: ".config/agents",
    ruleFile: "AGENTS.md",
  },
  {
    id: "openclaw",
    name: "OpenClaw",
    defaultEnabled: false,
    rootDirSuffix: ".openclaw",
    ruleFile: "AGENTS.md",
  },
  {
    id: "qoder",
    name: "Qoder",
    defaultEnabled: false,
    rootDirSuffix: ".qoder",
    ruleFile: "AGENTS.md",
  },
  {
    id: "codebuddy",
    name: "CodeBuddy",
    defaultEnabled: false,
    rootDirSuffix: ".codebuddy",
    ruleFile: "AGENTS.md",
  },
];

export const DEFAULT_ENABLED_AGENT_PRESET_IDS = AGENT_PRESETS.filter((preset) => preset.defaultEnabled).map(
  (preset) => preset.id,
);

const AGENT_PRESET_BY_ID = new Map(AGENT_PRESETS.map((preset) => [preset.id, preset]));

export function normalizeAgentPresetId(value: string): string {
  return value.trim().toLowerCase();
}

export function isBuiltInAgentPreset(agentType: string): boolean {
  return AGENT_PRESET_BY_ID.has(normalizeAgentPresetId(agentType));
}

export function getAgentPresetById(agentType: string): AgentPreset | undefined {
  return AGENT_PRESET_BY_ID.get(normalizeAgentPresetId(agentType));
}

export function resolveAgentPresetRootDir(homePath: string, agentType: string): string {
  const preset = getAgentPresetById(agentType);
  if (!preset) {
    return "";
  }
  const normalizedHome = homePath.trim().replace(/[\\/]+$/, "");
  if (!normalizedHome) {
    return "";
  }
  return `${normalizedHome}/${preset.rootDirSuffix}`;
}

export function resolveAgentPresetRuleFile(agentType: string): string {
  return getAgentPresetById(agentType)?.ruleFile ?? "AGENTS.md";
}

export function toAgentPresetSortWeight(agentType: string): number {
  const normalized = normalizeAgentPresetId(agentType);
  const index = AGENT_PRESETS.findIndex((preset) => preset.id === normalized);
  if (index >= 0) {
    return index;
  }
  return AGENT_PRESETS.length;
}

File: src-tauri/src/control_plane/agent_presets.rs
use std::path::PathBuf;

#[derive(Debug, Clone, Copy)]
pub struct AgentPreset {
    pub id: &'static str,
    pub home_suffix: &'static str,
    pub rule_file: &'static str,
    pub default_enabled: bool,
}

pub const BUILTIN_AGENT_PRESETS: [AgentPreset; 14] = [
    AgentPreset {
        id: "codex",
        home_suffix: ".codex",
        rule_file: "AGENTS.md",
        default_enabled: true,
    },
    AgentPreset {
        id: "claude",
        home_suffix: ".claude",
        rule_file: "CLAUDE.md",
        default_enabled: true,
    },
    AgentPreset {
        id: "gemini",
        home_suffix: ".gemini",
        rule_file: "AGENTS.md",
        default_enabled: true,
    },
    AgentPreset {
        id: "copilot",
        home_suffix: ".copilot",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
    AgentPreset {
        id: "cursor",
        home_suffix: ".cursor",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
    AgentPreset {
        id: "windsurf",
        home_suffix: ".codeium/windsurf",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
    AgentPreset {
        id: "kiro",
        home_suffix: ".kiro",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
    AgentPreset {
        id: "trae",
        home_suffix: ".trae",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
    AgentPreset {
        id: "opencode",
        home_suffix: ".config/opencode",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
    AgentPreset {
        id: "roo",
        home_suffix: ".roo",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
    AgentPreset {
        id: "amp",
        home_suffix: ".config/agents",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
    AgentPreset {
        id: "openclaw",
        home_suffix: ".openclaw",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
    AgentPreset {
        id: "qoder",
        home_suffix: ".qoder",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
    AgentPreset {
        id: "codebuddy",
        home_suffix: ".codebuddy",
        rule_file: "AGENTS.md",
        default_enabled: false,
    },
];

pub fn all_builtin_agent_presets() -> &'static [AgentPreset] {
    &BUILTIN_AGENT_PRESETS
}

pub fn find_builtin_agent_preset(agent_type: &str) -> Option<&'static AgentPreset> {
    let normalized = agent_type.trim().to_lowercase();
    BUILTIN_AGENT_PRESETS
        .iter()
        .find(|preset| preset.id == normalized)
}

pub fn is_builtin_agent_preset(agent_type: &str) -> bool {
    find_builtin_agent_preset(agent_type).is_some()
}

pub fn default_agent_enabled(agent_type: &str) -> bool {
    find_builtin_agent_preset(agent_type)
        .map(|preset| preset.default_enabled)
        .unwrap_or(false)
}

pub fn default_agent_rule_file(agent_type: &str) -> String {
    find_builtin_agent_preset(agent_type)
        .map(|preset| preset.rule_file.to_string())
        .unwrap_or_else(|| "AGENTS.md".to_string())
}

pub fn default_agent_root_dir(agent_type: &str) -> String {
    let Some(preset) = find_builtin_agent_preset(agent_type) else {
        return String::new();
    };

    let Some(home) = dirs::home_dir() else {
        return String::new();
    };

    let suffix = preset.home_suffix;
    if suffix.is_empty() {
        return String::new();
    }

    let path = PathBuf::from(home).join(suffix);
    path.to_string_lossy().to_string()
}
